### PR TITLE
Abiv2 use bytes string longer than 32 bytes

### LINKED ISSUE #7180: [Abiv2 proto fuzzer] Use variable length literals to initialize arguments of `bytes`/s
## Abstract

In #7177, after taking a look at  @ekpyron  (see https://github.com/ethereum/solidity/pull/7177#discussion_r310948711), I feel that the current approach that the proto fuzzer takes to initialize arguments of type `bytes`/`string` is insufficient to test abi coding. 

Currently, the proto fuzzer always initializes a `bytes` argument with a value that is exactly 32 bytes in size.

Ideally, we should be using a variable length literal to initialize an argument of `bytes`/`string` type.

## Motivation

Presently, the routine that provides the literal used to initialize a `bytes` or a `string` type variable is this

https://github.com/ethereum/solidity/blob/cd563e526a92721a1e9e1633f5e2c9e53e77b242/test/tools/ossfuzz/protoToAbiV2.h#L242-L245

The routine hashes a monotonically increasing counter using  `keccak256` and assigns  the resulting 32 byte hash value (in hex encoding minus the `0x` prefix) to a variable of `bytes` or `string` type. The initialized value is a string literal (that holds the hash value).

TBD.

### PR DESCRIPTION
**Pull request by bshastry**
fixes #7180

### CODE REVIEW

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 1]
```diff
@@ -204,12 +204,24 @@ std::string ProtoConverter::addressValueAsString(unsigned _counter)
 		.render();
 }
 
-std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _counter)
+/// Returns a hex literal if _isHexLiteral is true, a string literal otherwise.
+/// The flag _encloseInDoubleQuotes is true by default. It is false only when
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Bug: If `_isHexLiteral` is false it returns a string that contains `_width*2` bytes. It should return `_width` bytes.

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 2]
```diff
@@ -220,16 +229,113 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
 	unsigned startPos = 66 - numMaskNibbles;
+	// Extracts the least significant numMaskNibbles from the result
+	// of maskUnsignedIntToHex().
+	return maskUnsignedIntToHex(
+		_counter,
+		numMaskNibbles
+	).substr(startPos, numMaskNibbles);
+}
+
+/// Returns a hex literal if _isHexLiteral is true, a string literal otherwise.
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Class member functions should be documented in the header. Implementation-specific documentation goes into the function body, interface documentation goes to the header.

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 3]
```diff
@@ -220,16 +229,108 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
+			.render();
+
+	// Since we create a larger literal by duplicating the first
+	// 32 bytes, we can cache it. This saves us expensive calls
+	// to keccak256.
+	unsigned batchSize = _numBytes >= 32 ? 32 : _numBytes;
+	std::string cachedString = hexValueAsString(
+		batchSize,
+		_counter,
+		/*isHexLiteral=*/_isHexLiteral,
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Do we actually use such comments?

File: test/tools/ossfuzz/protoToAbiV2.h [hunk 4]
```diff
@@ -288,6 +281,22 @@ class ProtoConverter
 	static std::string integerValueAsString(bool _sign, unsigned _width, unsigned _counter);
 	static std::string addressValueAsString(unsigned _counter);
 	static std::string fixedByteValueAsString(unsigned _width, unsigned _counter);
+	/// Returns a hex literal if _isHexLiteral is true, a string literal otherwise.
+	/// The size of the returned literal is _numBytes bytes.
+	/// The flag _decorate is true by default.
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.h]: Maybe explain what `_decorate` is used for instead?

File: test/tools/ossfuzz/protoToAbiV2.h [hunk 5]
```diff
@@ -389,6 +399,28 @@ class ProtoConverter
 		);
 	}
 
+	/// Accepts a monotonically increasing counter as input and
+	/// returns a value to be used to compute the length of a
+	/// hex or string literal. This function is called by
+	/// bytesArrayValueAsString().
+	static unsigned getVarLength(unsigned _counter)
+	{
+		return (_counter + 879) * 32 % (s_maxDynArrayLength + 1);
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.h:435]: All these magic numbers are a mistery to me ;)

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 6]
```diff
@@ -730,8 +841,12 @@ void ProtoConverter::writeHelperFunctions()
 		internal pure returns (bytes memory, bytes memory)
 	{
 		bytes memory validEncoding = new bytes(4 + argumentEncoding.length);
-		// Ensure that invalidEncoding crops at least one and at most all bytes from correct encoding.
-		uint invalidLength = invalidLengthFuzz % argumentEncoding.length;
+		// Ensure that invalidEncoding crops at least 32 bytes (padding length
+		// is at most 31 bytes) since shorter bytes/string values can lead to
```
Comment (ekpyron) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: We could actually do this conditionally, i.e. only if the last argument of the function is in fact ``bytes`` or ``string``... Would that information be easily available here? If not it's not too important.

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 3 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: Actually, I'm not sure if this inline comment serves any purpose.

File: test/tools/ossfuzz/protoToAbiV2.h (hunk 5 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.h:435]: Yeah, this is hacky. I should have left a comment there.

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 6 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: Okay, I can do that.

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 3 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: But how about in a call like so

```
string x= hexValueAsString(size, counter, true, false);
```

I think an inline comment in such calls make the intent explicit.

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 6 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: I'm wondering if we would need to crop at least 32 bytes if the last argument to the function is an array. May be I should simply read the abi encoder specification :-)

Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: For reference, here is what the encoder spec says

> uint<M>: enc(X) is the big-endian encoding of X, padded on the higher-order (left) side with zero-bytes such that the length is 32 bytes.

> int<M>: enc(X) is the big-endian two’s complement encoding of X, padded on the higher-order (left) side with 0xff for negative X and with zero bytes for positive X such that the length is 32 bytes.

So doesn't that mean we may have to crop at least 31 bytes for (u)intX as well in order to guarantee an invalid encoding?

Also, since encoding of an array are composed of encoding of the base type, doesn't it mean we would need to extend the cropping strategy to static/dynamic
  - arrays of bytes/string?
  - arrays of (u)intX?

Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: Ah, just realized my mistake. (u)intX are left padded. Only bytes/string are right padded. So if I understood correctly, we should conditionally chop off at least 32 bytes when the last argument to the function is one of the following types:
  - bytes/string
  - base type of a static/dynamic array is bytes/string

**Comment by bshastry** [comment]
Updated PR. I assumed that if the type of last parameter is one of the following, then at least 32 bytes need to be cropped from a valid abi encoding to create an invalid encoding
  - bytes/string
  - arrays whose base type is bytes/string

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 7]
```diff
@@ -63,6 +63,8 @@ void ProtoConverter::visitType(
 	std::string varName, paramName;
 	createDeclAndParamList(_type, _dataType, varName, paramName);
 	addCheckedVarDef(_dataType, varName, paramName, _value);
+	// Update right paddedness of type
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]:
```suggestion
	// Update right padding of type
```

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 8]
```diff
@@ -695,20 +810,34 @@ void ProtoConverter::visit(TestFunction const& _x)
...
+			<invalidLengthFuzz>,
+			<isRightPadded>
+		);
 		if (returnVal != 0)
 			return uint(200000) + returnVal;
+		</atLeastOneVar>
 		return 0;
 	}
 	)")
 	("parameter_names", dev::suffixedVariableNameList(s_varNamePrefix, 0, m_varCounter))
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: In general, the naming seems to be inconsistent here. Or is this desired?

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 9]
```diff
@@ -220,16 +231,108 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
+}
+
+std::string ProtoConverter::hexValueAsString(
+	unsigned _numBytes,
+	unsigned _counter,
+	bool _isHexLiteral,
+	bool _decorate
+)
+{
+	solAssert(_numBytes <= 32, "Proto ABIv2 fuzzer: Invalid hex length");
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: As far as I understand I can't be `0` because of the assertion in `variableLengthValueAsString` right? So maybe also add it here for clarity?

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 10]
```diff
@@ -220,16 +231,108 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
+	// 32 bytes, we can cache it. This saves us expensive calls
+	// to keccak256.
+	unsigned batchSize = _numBytes >= 32 ? 32 : _numBytes;
+	std::string cachedString = hexValueAsString(
+		batchSize,
+		_counter,
+		_isHexLiteral,
+		/*decorate=*/false
+	);
+	_numBytes -= batchSize;
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Maybe add a new variable here? I find it a bit hard to follow `_numBytes` through the whole function. Actually, could it be refactored into separate functions? Please see my comments below :)

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 11]
```diff
@@ -220,16 +231,108 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
+	std::string output = cachedString;
+	// Check if size of cached string meets expectations
+	bool sanityCheck = false;
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Could add a function `checkCachedString` or sth. like that and have the assertion in there.

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 12]
```diff
@@ -220,16 +231,108 @@ std::string ProtoConverter::fixedByteValueAsString(unsigned _width, unsigned _co
...
+			cachedString.size() == 32;
+
+	solAssert(
+		_numBytes == 0 || sanityCheck,
+		"Proto ABIv2 fuzzer: Logic flaw in variable literal creation"
+	);
+
+	// Append cached literal until we are able to produce an
+	// output string that is _numBytes in size.
+	while (_numBytes > 0)
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp]: Other than the initial assignment to `output`, it's only used here again. Isn't it? So maybe add a function call `output = concatOutput(output, cachedString, numBytes ...)`

File: test/tools/ossfuzz/protoToAbiV2.h [hunk 13]
```diff
@@ -405,8 +466,13 @@ class ProtoConverter
 	unsigned m_varCounter;
 	/// Monotonically increasing return value for error reporting
 	unsigned m_returnValue;
+	/// Flag to check if last parameter passed of a function call
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.h]:
```suggestion
	/// Flag that indicates if last parameter passed to a function call
```

File: test/tools/ossfuzz/protoToAbiV2.h [hunk 14]
```diff
@@ -389,6 +423,33 @@ class ProtoConverter
...
+	static unsigned getVarLength(unsigned _counter)
+	{
+		// Since _counter values are usually small, we use
+		// this linear equation to make the number derived from
+		// _counter approach a uniform distribution over
+		// [0, s_maxDynArrayLength]
+		return (_counter + 879) * 32 % (s_maxDynArrayLength + 1);
+	}
+
+	/// Returns a variable length hex/string literal whose value and size are pseudo-
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.h]:
```suggestion
	/// Returns a hex/string literal of variable length whose value and size are pseudo-
```

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 15]
```diff
@@ -724,14 +853,31 @@ void ProtoConverter::writeHelperFunctions()
...
+		bytes4 funcSelector,
+		bytes memory argumentEncoding,
+		uint invalidLengthFuzz,
+		bool isRightPadded
+	) internal pure returns (bytes memory, bytes memory)
 	{
 		bytes memory validEncoding = new bytes(4 + argumentEncoding.length);
-		// Ensure that invalidEncoding crops at least one and at most all bytes from correct encoding.
-		uint invalidLength = invalidLengthFuzz % argumentEncoding.length;
+		// Ensure that invalidEncoding crops at least 32 bytes (padding length
```
Comment (erak) [inline test/tools/ossfuzz/protoToAbiV2.cpp:892]: Could this also be done in `ProtoConverter::visit(TestFunction)` such that the invalid length is passed directly? Or generally speaking, do we prefer to have as much code as possible on the C++ or on the Solidity side?

I also think that the `ProtoConverter` could be eventually split up such that Solidity "codegen", that's the string creation. But that's not in scope for this PR, I'd guess :)

**Review by erak — CHANGES_REQUESTED** [review]
I have a few minor comments with regards to the code structure. But in general it looks good!

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 8 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: No, it was my oversight. Will rename it to `parameterNames`

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 15 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp:892]: > Could this also be done in ProtoConverter::visit(TestFunction) such that the invalid length is passed directly?

Invalid length is first passed (in C++ code) here: https://github.com/ethereum/solidity/pull/7188/files#diff-7da1fb83eed279ed546eb9cb40618f9aR838

Since the C++ portion does not yet know what the size of the valid encoding is (we need to know the size of the valid encoding to place an upper bound on the length of the invalid encoding), we need to adjust the value passed from C++ inside solidity code.

Valid encoding size is computed in Solidity code by first making a call to abi encode on the parameter list here
https://github.com/ethereum/solidity/pull/7188/files#diff-7da1fb83eed279ed546eb9cb40618f9aR814

...and then adding the function selector size (4 bytes) here
https://github.com/ethereum/solidity/pull/7188/files#diff-7da1fb83eed279ed546eb9cb40618f9aR866

Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp:892]: > I also think that the ProtoConverter could be eventually split up such that Solidity "codegen", that's the string creation. But that's not in scope for this PR, I'd guess :)

Yeah, I will create an issue for this :-)

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 10 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: Sure, I'll take a look at this snippet again

**Comment by erak** [comment]
Do you want to take my suggestions regarding the tiny refactoring of `variableLengthValueAsString` into account? I think it's also ready to be merged without.

**Comment by bshastry** [comment]
> Do you want to take my suggestions regarding the tiny refactoring of `variableLengthValueAsString` into account? I think it's also ready to be merged without.

Sorry for the delay on this. Actually, I was waiting for @ekpyron feedback on the implementation of the padding check. Will take a second look at @erak refactoring comment soon. Context switching is making me take longer to come up with an elegant refactoring :-)

**Comment by ekpyron** [comment]
Just had a quick look at the padding check and that looks good to me in any case - let me know, if you'll refactor this some more or not, then I'll have another more detailed look.

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 6 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: @ekpyron Is my assessment here correct? Basically, the addition to what you requested above (https://github.com/ethereum/solidity/pull/7188#discussion_r312000815) is that we crop at least 32 bytes for arrays whose base type is bytes/string.

Comment (ekpyron) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: I think so - if the last argument is ``bytes`` or ``string`` itself or an array with one of these as last base type, then encodings without padding at the end are fine, so that's the cases in which we have to crop at least 32 bytes. I think that's fine!

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 10 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp]: Refactoring was absolutely required :) hope the code is more readable now!

**Comment by bshastry** [comment]
@erak Updated PR to mainly address your refactoring comment.

**Review by erak — APPROVED** [review]

File: test/tools/ossfuzz/protoToAbiV2.cpp (hunk 15 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToAbiV2.cpp:892]: @erak I was going to create an issue for this but realized I didn't follow your comment. Do you mind briefly commenting what you had in mind?
